File: packages/flutter/lib/models/json_model.dart
/// Shared helpers for decoding SDK JSON model maps.
///
/// The SDK model layer intentionally remains thin: it accepts contract-owned
/// camelCase/snake_case aliases, preserves unknown adapter fields, and lets the
/// core/contract own business semantics and validation defaults.
library;

/// Returns a shallow copy that model instances can keep as their raw payload.
Map<String, dynamic> jsonModelRaw(Map<String, dynamic> map) =>
    Map<String, dynamic>.from(map);

/// Merges a model's canonical encoded fields over its raw payload.
///
/// Unknown fields survive round-trips while normalized fields use the SDK's
/// canonical camelCase representation.
Map<String, dynamic> mergePreservedJsonModel(
  Map<String, dynamic> raw,
  Map<String, dynamic> canonical,
) {
  final merged = Map<String, dynamic>.from(raw);
  merged.addAll(canonical);
  return merged;
}

DateTime? jsonDateTimeField(Map<String, dynamic> map, List<String> keys) {
  for (final key in keys) {
    final value = map[key];
    if (value is String) return DateTime.tryParse(value);
  }
  return null;
}

int jsonIntField(
  Map<String, dynamic> map,
  List<String> keys, {
  int defaultValue = 0,
}) {
  for (final key in keys) {
    final value = map[key];
    if (value is int) return value;
    if (value is num) return value.toInt();
  }
  return defaultValue;
}

double jsonDoubleField(
  Map<String, dynamic> map,
  List<String> keys, {
  double defaultValue = 0,
}) {
  for (final key in keys) {
    final value = map[key];
    if (value is num) return value.toDouble();
  }
  return defaultValue;
}

bool jsonBoolField(
  Map<String, dynamic> map,
  List<String> keys, {
  bool defaultValue = false,
}) {
  for (final key in keys) {
    final value = map[key];
    if (value is bool) return value;
  }
  return defaultValue;
}

String? jsonStringField(
  Map<String, dynamic> map,
  List<String> keys, {
  bool nonEmpty = true,
}) {
  for (final key in keys) {
    final value = map[key];
    if (value is String && (!nonEmpty || value.isNotEmpty)) return value;
  }
  return null;
}

List<String> jsonStringListField(Map<String, dynamic> map, List<String> keys) {
  for (final key in keys) {
    final value = map[key];
    if (value is List) return value.map((item) => item.toString()).toList();
  }
  return const [];
}

List<T> jsonObjectListField<T>(
  Object? value,
  T Function(Map<String, dynamic>) convert,
) {
  if (value is! List) return const [];
  return value
      .whereType<Map>()
      .map((item) => convert(Map<String, dynamic>.from(item)))
      .toList();
}

File: packages/ios/Sources/Napaxi/JsonModel.swift
import Foundation

// iOS adapter parity for the Flutter shared JSON model helpers
// (`packages/flutter/lib/models/json_model.dart`). The Flutter model layer is
// intentionally thin: it accepts contract-owned camelCase/snake_case aliases,
// preserves unknown adapter fields, and lets core/contract own business
// semantics. These free functions are the Swift equivalents over
// `NapaxiJSONValue` maps, so each adapter decodes model maps the same way.

/// Returns a shallow copy that model instances can keep as their raw payload.
public func jsonModelRaw(_ map: [String: NapaxiJSONValue]) -> [String: NapaxiJSONValue] {
    map
}

/// Merges a model's canonical encoded fields over its raw payload. Unknown
/// fields survive round-trips while normalized fields use the canonical
/// representation.
public func mergePreservedJsonModel(
    _ raw: [String: NapaxiJSONValue],
    _ canonical: [String: NapaxiJSONValue]
) -> [String: NapaxiJSONValue] {
    var merged = raw
    for (key, value) in canonical {
        merged[key] = value
    }
    return merged
}

/// Decode the first present key as an ISO-8601 / RFC-3339 date string.
public func jsonDateTimeField(_ map: [String: NapaxiJSONValue], _ keys: [String]) -> Date? {
    for key in keys {
        if let value = map[key]?.stringValue, let date = napaxiParseDate(value) {
            return date
        }
    }
    return nil
}

/// Decode the first present key as an integer.
public func jsonIntField(
    _ map: [String: NapaxiJSONValue],
    _ keys: [String],
    defaultValue: Int = 0
) -> Int {
    for key in keys {
        if let number = map[key]?.numberValue {
            return Int(number)
        }
    }
    return defaultValue
}

/// Decode the first present key as a double.
public func jsonDoubleField(
    _ map: [String: NapaxiJSONValue],
    _ keys: [String],
    defaultValue: Double = 0
) -> Double {
    for key in keys {
        if let number = map[key]?.numberValue {
            return number
        }
    }
    return defaultValue
}

/// Decode the first present key as a bool.
public func jsonBoolField(
    _ map: [String: NapaxiJSONValue],
    _ keys: [String],
    defaultValue: Bool = false
) -> Bool {
    for key in keys {
        if let value = map[key]?.boolValue {
            return value
        }
    }
    return defaultValue
}

/// Decode the first present key as a string. When `nonEmpty` is true, empty
/// strings are skipped.
public func jsonStringField(
    _ map: [String: NapaxiJSONValue],
    _ keys: [String],
    nonEmpty: Bool = true
) -> String? {
    for key in keys {
        if let value = map[key]?.stringValue, !nonEmpty || !value.isEmpty {
            return value
        }
    }
    return nil
}

/// Decode the first present key as a list of strings.
public func jsonStringListField(
    _ map: [String: NapaxiJSONValue],
    _ keys: [String]
) -> [String] {
    for key in keys {
        if let array = map[key]?.arrayValue {
            return array.map { $0.stringValue ?? napaxiStringify($0) }
        }
    }
    return []
}

private func napaxiParseDate(_ value: String) -> Date? {
    let iso = ISO8601DateFormatter()
    iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = iso.date(from: value) { return date }
    iso.formatOptions = [.withInternetDateTime]
    return iso.date(from: value)
}

private func napaxiStringify(_ value: NapaxiJSONValue) -> String {
    switch value {
    case .string(let string): return string
    case .number(let number):
        return number.rounded() == number ? String(Int(number)) : String(number)
    case .bool(let bool): return String(bool)
    case .null: return ""
    case .array, .object: return (try? NapaxiRawJSON(value).jsonString()) ?? ""
    }
}
